components/server/src/api/server.ts

```diff
@@ -217,50 +217,63 @@ export class API {
+                    // actually call the RPC handler
+                    const auth = async () => {
+                        // Authenticate
+                        const userId = await self.verify(connectContext);
+                        const isAuthenticated = !!userId;
                         const requiresAuthentication = !Unauthenticated.get(target, prop);
-
                         if (!isAuthenticated && requiresAuthentication) {
                             throw new ConnectError("unauthenticated", Code.Unauthenticated);
                         }
 
                         if (isAuthenticated) {
-                            await rateLimit(subjectId);
-                            context.user = await self.ensureFgaMigration(subjectId);
+                            await rateLimit(userId);
+                            await self.ensureFgaMigration(userId);
                         }
-
                         // TODO(at) if unauthenticated, we still need to apply enforece a rate limit
 
-                        // actually call the RPC handler
+                        return userId ? SubjectId.fromUserId(userId) : undefined;
+                    };
+
+                    const apply = async <T>(): Promise<T> => {
                         return Reflect.apply(target[prop as any], target, args);
                     };
                     if (grpc_type === "unary" || grpc_type === "client_stream") {
                         return withRequestContext(async () => {
                             try {
-                                const promise = await apply<Promise<any>>();
-                                const result = await promise;
+                                const subjectId = await auth();
+                                const result = await runWithSubjectId(subjectId, async () => {
+                                    const promise = await apply<Promise<any>>();
+                                    return await promise;
+                                });
                                 done();
                                 return result;
                             } catch (e) {
                                 handleError(e);
                             }
                         });
                     }
+
+                    // Because we can't await before returning that generator, we await inside the generator, and create child contexts with that SubjectId
                     return wrapAsyncGenerator(
                         (async function* () {
                             try {
-                                const generator = await apply<AsyncGenerator<any>>();
-                                for await (const item of generator) {
-                                    yield item;
+                                const subjectId = await auth();
+                                const generator = await runWithSubjectId(subjectId, () => apply<AsyncGenerator<any>>());
```

Comment:
@akosyakov because I don't know how the resulting state machine looks like, I'm not entirely sure if we need to wrap this in `runWithSubjectId`, but I think we do.

Reply:
that's fine I think :D